Incremental build failed

Steps to reproduce:
1. build a package -> success
2. add error in sources, rebuild package -> BUILD step fails
3. revert changes to sources, rebuild -> BUILD step is *not* executed -> FAIL

The problem is that Bob does not invalidate the result and input of the build step prior to building it.

But the package step isn't executed in 3?

```bash
binary-repository $ bob dev root
>> root
>> root/lib
   DOWNLOAD  dev/dist/lib/1/workspace...[Errno 111] Connection refused
   CHECKOUT  dev/src/lib/1/workspace
   BUILD     dev/build/lib/1/workspace
   PACKAGE   dev/dist/lib/1/workspace
>> root
   CHECKOUT  dev/src/root/1/workspace
   BUILD     dev/build/root/1/workspace
   PACKAGE   dev/dist/root/1/workspace
Build result is in dev/dist/root/1/workspace
```
-> edit dev/src/lib/1/workspace/lib.c

```bash
binary-repository $ bob dev root>> root
>> root/lib
   CHECKOUT  skipped (fixed package dev/src/lib/1/workspace)
   BUILD     dev/build/lib/1/workspace
/home/biera/projects/bob/bob-tutorials/binary-repository/dev/src/lib/1/workspace/lib.c: In function 'foo':
/home/biera/projects/bob/bob-tutorials/binary-repository/dev/src/lib/1/workspace/lib.c:5:2: error: unknown type name 'p'
  p rintf("Hello world!\n");
  ^
/home/biera/projects/bob/bob-tutorials/binary-repository/dev/src/lib/1/workspace/lib.c:5:10: error: expected declaration specifiers or '...' before string constant
  p rintf("Hello world!\n");
          ^
Step failed on line 17: Exit status 1; Command: gcc -c -o lib.o $1/lib.c
Build error: Build script ./dev/build/lib/1/build.sh returned with 1
Processing stack: root/lib
You may resume at this point with '--resume' after fixing the error.
```

revert

```bash
binary-repository $ bob dev root
>> root
>> root/lib
   CHECKOUT  skipped (fixed package dev/src/lib/1/workspace)
   BUILD     skipped (unchanged input for dev/build/lib/1/workspace)
   PACKAGE   skipped (unchanged input for dev/dist/lib/1/workspace)
>> root
   CHECKOUT  skipped (fixed package dev/src/root/1/workspace)
   BUILD     skipped (unchanged input for dev/build/root/1/workspace)
   PACKAGE   skipped (unchanged input for dev/dist/root/1/workspace)
Build result is in dev/dist/root/1/workspace
```

I got a repro case though it was certainly not the one I had today:

1. build
```
$ bob dev root
>> root
>> root/lib
   DOWNLOAD  dev/dist/lib/1/workspace...[Errno 111] Connection refused
   CHECKOUT  dev/src/lib/1/workspace
   BUILD     dev/build/lib/1/workspace
   PACKAGE   dev/dist/lib/1/workspace
>> root
   CHECKOUT  dev/src/root/1/workspace
   BUILD     dev/build/root/1/workspace
   PACKAGE   dev/dist/root/1/workspace
Build result is in dev/dist/root/1/workspace
```
2. insert error in lib.c and rebuild clean:
```
$ bob dev root --clean
>> root
>> root/lib
   CHECKOUT  skipped (fixed package dev/src/lib/1/workspace)
   BUILD     dev/build/lib/1/workspace
/home/jkloetzke/src/bob/bob-tutorials/binary-repository/dev/src/lib/1/workspace/lib.c: In function 'foo':
/home/jkloetzke/src/bob/bob-tutorials/binary-repository/dev/src/lib/1/workspace/lib.c:5:2: error: unknown type name 'p'
  p rintf("Hello world!\n");
  ^
/home/jkloetzke/src/bob/bob-tutorials/binary-repository/dev/src/lib/1/workspace/lib.c:5:10: error: expected declaration specifiers or '...' before string constant
  p rintf("Hello world!\n");
          ^~~~~~~~~~~~~~~~
Step failed on line 17: Exit status 1; Command: gcc -c -o lib.o $1/lib.c
Build error: Build script ./dev/build/lib/1/build.sh returned with 1
Processing stack: root/lib
You may resume at this point with '--resume' after fixing the error.
```
3. revert and change lib.yaml `packageScript` to force re-execution of package step:
```
$ bob dev root --clean
>> root
   PRUNE     dev/dist/root/1/workspace (recipe changed)
>> root/lib
   PRUNE     dev/dist/lib/1/workspace (recipe changed)
   DOWNLOAD  dev/dist/lib/1/workspace...[Errno 111] Connection refused
   CHECKOUT  skipped (fixed package dev/src/lib/1/workspace)
   BUILD     skipped (unchanged input for dev/build/lib/1/workspace)
   PACKAGE   dev/dist/lib/1/workspace
cp: cannot stat '/home/jkloetzke/src/bob/bob-tutorials/binary-repository/dev/build/lib/1/workspace/lib.o': No such file or directory
Step failed on line 17: Exit status 1; Command: cp -a $1/lib.o .
Build error: Build script ./dev/dist/lib/1/package.sh returned with 1
Processing stack: root/lib
```

--- a/pym/bob/cmds/build.py
+++ b/pym/bob/cmds/build.py
@@ -576,6 +576,14 @@ esac
                     BobState().setDirectoryState(prettySrcPath,
                         { d:s for (d,s) in checkoutState.items() if d is not None })
 
+                    # Forge checkout result before we run the step again.
+                    # Normally the correct result is set directly after the
+                    # checkout finished. But if the step fails and the user
+                    # re-runs with "build-only" the dependent steps should
+                    # trigger.
+                    if BobState().getResultHash(prettySrcPath) is not None:
+                        BobState().setResultHash(prettySrcPath, datetime.datetime.utcnow())
+
                     print(colorize("   CHECKOUT  {}".format(prettySrcPath), "32"))
                     self._runShell(checkoutStep, "checkout")
 
@@ -634,6 +642,12 @@ esac
             else:
                 print(colorize("   BUILD     {}".format(prettyBuildPath), "32"))
                 if self.__cleanBuild: emptyDirectory(prettyBuildPath)
+                # Squash state because running the step will change the
+                # content. If the execution fails we have nothing reliable
+                # left and we _must_ run it again.
+                BobState().delInputHashes(prettyBuildPath)
+                BobState().setResultHash(prettyBuildPath, datetime.datetime.utcnow())
+                # build it
                 self._runShell(buildStep, "build")
                 # Use timestamp in release mode and only hash in development mode
                 BobState().setResultHash(prettyBuildPath,
@@ -719,6 +733,9 @@ esac
                 else:
                     print(colorize("   PACKAGE   {}".format(prettyPackagePath), "32"))
                     emptyDirectory(prettyPackagePath)
+                    # invalidate result because folder was cleared
+                    BobState().delInputHashes(prettyPackagePath)
+                    BobState().setResultHash(prettyPackagePath, datetime.datetime.utcnow())
                     self._runShell(packageStep, "package")
                     packageExecuted = True
                     if packageBuildId and self.__archive.canUploadLocal():

--- a/test/rerun-failed-build/recipes/foo.c
+++ b/test/rerun-failed-build/recipes/foo.c
@@ -1,0 +1,7 @@
+#include <stdio.h>
+
+int main(void)
+{
+	printf("Hello world!\n");
+	return 0;
+}

--- a/test/rerun-failed-build/recipes/root.yaml
+++ b/test/rerun-failed-build/recipes/root.yaml
@@ -1,0 +1,12 @@
+root: True
+
+checkoutScript: |
+    [ -e foo.c ] || cp $<<foo.c>> foo.c
+
+buildScript: |
+    rm -rf *
+    gcc -o foo $1/foo.c
+
+packageVars: [FOO]
+packageScript: |
+    cp -a $1/foo .

--- a/test/rerun-failed-build/run.sh
+++ b/test/rerun-failed-build/run.sh
@@ -1,0 +1,20 @@
+#!/bin/bash -e
+. ../test-lib.sh 2>/dev/null || { echo "Must run in script directory!" ; exit 1 ; }
+cleanup
+
+# If a step is run Bob must not rely on the old state of the workspace anymore.
+# Failing steps must thus always be run again.
+
+# build normally
+run_bob dev root
+
+# introduce some error in the sources -> expected fail
+echo "#error fail" > dev/src/root/1/workspace/foo.c
+if run_bob dev root 2>/dev/null ; then
+	echo "Expected fail"
+	exit 1
+fi
+
+# run again and force package step
+rm dev/src/root/1/workspace/foo.c
+run_bob dev root -DFOO=bar

